config/v1/types_ingress.go

```diff
@@ -41,13 +43,122 @@ type IngressSpec struct {
 	// certificate.
 	// +optional
 	AppsDomain string `json:"appsDomain,omitempty"`
+
+	// componentRoutes is an optional list of routes configured by the cluster-admin to
+	// override the default hostname and serving certificates. The namespace and name of
+	// each componentRoute should match an existing componentRoute defined in the status.
+	//
+	// To determine the set of configurable Routes, look at namespace and name of entries in the
+	// .status.componentRoutes, where participating operators place current route status.
+	// +optional
+	ComponentRoutes []ComponentRouteSpec `json:"componentRoutes,omitempty"`
 }
 
+// consumingUser is an alias for string which we add validation to. Currently only service accounts are supported.
+// +kubebuilder:validation:Pattern="^system:serviceaccount:[a-z0-9]([-a-z0-9]*[a-z0-9])?:[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=512
+type consumingUser string
+
+// hostname is an alias for hostname string validation.
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=253
+// +kubebuilder:validation:Format=hostname
+type hostname string
+
 type IngressStatus struct {
+	// componentRoutes is where participating operators place the current route status for routes whose
+	// hostnames and serving certificates can be customized by the cluster-admin.
+	//
+	// An operator that creates entries in this slice should clean them up during removal.
+	// An operator must also handle the case of deleted status without churn.
+	// +optional
+	ComponentRoutes []ComponentRouteStatus `json:"componentRoutes,omitempty"`
 }
 
-// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// ComponentRouteSpec allows for configuration of a route's hostname and serving certificate.
+type ComponentRouteSpec struct {
+	// namespace is the namespace of the route to customize. It must be a real namespace. Using an actual namespace
+	// ensures that no two components will conflict and the same component can be installed multiple times.
+	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:Required
+	// +required
+	Namespace string `json:"namespace"`
 
+	// name is the logical name of the route to customize.
```

Comment:
Maybe copy the verbiage from the corresponding status field, namely state for emphasis that the value of `name` need not match the API object's name, because that is a weird and potentially confusing nuance.

Reply:
I believe we should target the verbiage for the spec.ComponentRoute.{namespace|name} for cluster admins and the verbiage in the status for Operators. I will do my best to make it clearer in the next update.

Reply:
The cluster admin has to go to the status to see what routes can be customized, and then go to the spec to make the change, so it's hard to say which one the cluster admin is going to run `oc explain` on, or look up in godoc references.